File: 06-TREES/lowest-common-ancestor-of-a-binary-search-tree.py
from TreeNode import TreeNode
from collections import deque


# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        p_path = [root]
        tempRoot = root
        while tempRoot and tempRoot.val != p.val:
            tempRoot = tempRoot.left if tempRoot.val > p.val else tempRoot.right
            p_path.append(tempRoot)

        q_path = [root]
        tempRoot = root
        while tempRoot and tempRoot.val != q.val:
            tempRoot = tempRoot.left if tempRoot.val > q.val else tempRoot.right
            q_path.append(tempRoot)

        for i in range(len(q_path) - 1, -1, -1):
            if q_path[i] in p_path:
                return q_path[i]

        return root

File: 06-TREES/PHP/lowest-common-ancestor-of-a-binary-search-tree.php
<?php

# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
class Solution
{
    public function lowestCommonAncestor(TreeNode $root, TreeNode $p, TreeNode $q): TreeNode
    {
        $p_path = [$root];
        $tempRoot = $root;
        while ($tempRoot and $tempRoot->val != $p->val) {
            $tempRoot = $tempRoot->val > $p->val ? $tempRoot->left : $tempRoot->right;
            $p_path[] = $tempRoot;
        }


        $q_path = [$root];
        $tempRoot = $root;
        while ($tempRoot and $tempRoot->val != $q->val) {
            $tempRoot = $tempRoot->val > $q->val ? $tempRoot->left : $tempRoot->right;
            $q_path[] = $tempRoot;
        }

        for ($i = count($q_path); $i >= 0; $i--) {
            if (in_array($q_path[$i], $p_path)) {
                return $p_path[$i];
            }
        }

        return $root;
    }
}